### PR TITLE
REF Add shape param to 2D and 3D

### PR DESCRIPTION
**Pull request by eickenberg**
This extends the refactoring by @lostanlen proposed in PR https://github.com/kymatio/kymatio/pull/195

It is probably better to comment over at that PR, in order to keep the conversation in one place.

If accepted, then this can be merged into @lostanlen s current PR

@janden @edouardoyallon @eugenium

### CODE REVIEW

**Comment by lostanlen** [comment]
Thanks @eickenberg. This is great. 
Could you describe your train of thoughts / proposed agenda on how this will evolve in the future?
What exactly do `J` and `shape` mean, and how do they relate to each other?

**Comment by eickenberg** [comment]
> What exactly do J and shape mean, and how do they relate to each other?

`J` is the number of scales (or whatever it has meant up until now)

`shape` is a tuple consisting of `M, N` or `M, N, O` and means what those mean now. That might be ambiguous. I was thinking it means signal shape - the shape of the incoming signal, in case this needs to be specified.

Moving forward, I would like to do some work on **dissociating** the scattering objects from their filterbanks again, in order to be able to pass a filter bank or a filter bank name at instantiation. If none of that is specified, do the default. I am mentioning this because I think that the shapes of the filters should be intrinsic to the filter bank. The shape provided as an argument to the scattering object should be compatible with the filter bank shape in some way, but it would also inform e.g. the cuda kernel size for performing the fast downsampling.

**Comment by lostanlen** [comment]
> `J` is the number of scales (or whatever it has meant up until now)
That definition is not true in 1D because of the `P` linearly spaced filters.

**Comment by eickenberg** [comment]
> That definition is not true in 1D because of the P linearly spaced filters.

Hence why I was hedging with what is in the parentheses.

If a filter bank is passed, then `J` can act as a cutoff maybe, or be overridden

**Comment by lostanlen** [comment]
What you are describing involves the creation of more objects, doesn't it?
In WaveletScattering.jl I experimented with this and had a lot of success.
I defined a `Spec` immutable object whose constructor only checks parameters and sets up defaults, then a `Bank` object which constructs the filter banks according to spec (and offers some amount of mutable behavior, like oversampling), and finally a `ScatteringLayer` object which is immutable. Check it out if you're interested. That was a lot of work to set up in Julia though, so doing it in Python would also be a lot of work (especially given that we don't have parametric inheritance nor immutability)

**Comment by lostanlen** [comment]
> If a filter bank is passed, then J can act as a cutoff maybe, or be overridden

Wait, so `J` can be changed at runtime now?

`J` is not very well defined at the moment in the API, except in the particular case `Q=1` and `slant=1`. But it's all over Mallat's papers so it's worth keeping it because it will be enough for 80% of our users.

**Comment by lostanlen** [comment]
> I don't see how slant affects the meaning of J. I think the 2D filters are always in octaves atm

The greater the slant, the longer the 2D wavelet in the orthogonal direction to its orientation
So the typical support of a 2D wavelet at scale `j` is `2**j` in the direction of its orientation and `slant*(2**j)` in the orthogonal direction.
Therefore, the receptive field size of the 2D scattering transform in Kymatio is `slant*2**J`

Whereas in 1D, the receptive field size is `2**J` whatever the value of `Q`.

In other words, in 1D `J` is used to determine a receptive field size, and the number of filters `P` is derived from it. Whereas in 2D `J` is used to determine a number of scales, and the receptive field size `slant*2**J` is derived from it.

This affects padding and the inference of filterbank `shape` from `J`
So I'm curious what you mean exactly by `J` and `shape`. Are they properties of the filter bank or of the receptive field size?

**Comment by lostanlen** [comment]
Ah, no, wait a minute, it's not that bad. `J` really *is* the number of octaves in Kymatio. Good.
There are no linearly spaced filters like in ScatNet, right?

**Comment by lostanlen** [comment]
Upon sleeping on it, I say we merge this. It does much much less harm than good.
Hey, we have a consistent API of the form `ScatteringNd(J,shape)`, it takes `N`-d signal of shape `shape` and returns a scattering transform sampled at `2**J`: *que demande le peuple* ?

**Comment by lostanlen** [comment]
@eickenberg why does this PR not rename `T` to `shape`?

**Comment by eickenberg** [comment]
It could - just didn't know whether that felt like the right name. It's
still consistent as is: second argument is a number or tuple indicating 1d
2d or 3d shape

**Comment by lostanlen** [comment]
hm, so you're saying it's consistent as positional arguments but not as kwarg?
is this a satisfying equilibrium? (maybe it is, I don't know)

**Comment by eickenberg** [comment]
I guess my hesitation was due to semantics: is "shape" a good word to
designate a length?

I'm fine with it, but I thought it might not be fine for everybody.

("Size" would fit better, but that word is ambiguously overloaded, since
numpy.ndarray.size means something other than torch.FloatTensor.size)

**Comment by lostanlen** [comment]
these questions could use some input from other Kymatio contributors, notably
@janden @edouardoyallon @eugenium @andreuxmath

**Comment by edouardoyallon** [comment]
Super! Merge?

**Comment by eugenium** [comment]
Well i guess this one would actually break any existing kymatio or pysctwave code... that being said i m prob the only one already using kymatio live so i guess ill be the only one to suffer

**Comment by eickenberg** [comment]
The point is to break it.

If we just change M, N, J to J, M, N, it might not break, but interpret M
as J, thereby asking for hundreds of scales :)

**Comment by lostanlen** [comment]
I wholeheartedly concur with @eickenberg. 1) We need to break API; and 2) better sooner than later.
The question is: how? :p

**Comment by eugenium** [comment]
Ah ok I see. in that case this seems fine

**Comment by lostanlen** [comment]
@janden do you have a thing to say before we merge this?

**Comment by janden** [comment]
I like `shape` for 1D too. It's better to be consistent here, and it's not _that_  odd to call the length a shape.

On the other hand, if we do change the name in 1D, I would vote for making integer values invalid and only allowing 1-tuples (i.e., `(T,)`).

Also,

> Ah, no, wait a minute, it's not that bad. `J` really _is_ the number of octaves in Kymatio. Good.

Unfortunately, `J` is not the number of octaves in Kymatio. Try setting `J=1` when `Q=8` and you'll see something interesting. Actually, this is a good thing, I think, since it decouples `J` and `Q` (unlike the approach in ScatNet).

> There are no linearly spaced filters like in ScatNet, right?

There are linearly spaced filters, however, but calculated slightly differently compared to ScatNet.

**Comment by lostanlen** [comment]
So what is `J` in 1-D? base-2 logarithm of receptive field size?

**Comment by eickenberg** [comment]
Alright, I'll make that change later on

**Comment by janden** [comment]
> Damn ... So `J` does not mean the same thing in `1D` and in `2D`.
> See this link: [#2 (comment)](https://github.com/lostanlen/kymatio/pull/2#issuecomment-441516407)

Yes unfortunately. This is a tricky one. One solution may be to fix `2**J` to be the size of the lowpass filter, period. Then the number of scales/wavelets/filters is adjusted so that they do not become wider than this. For 2D, increasing `slant` would then result in fewer octaves of wavelets for a fixed `2**J`.

**Comment by lostanlen** [comment]
> Against rejecting integers for 1D, though. I will make something that turns it into a tuple if it isn't one.

I agree. It would be good to have *some* type dispatch in the constructor.


> One solution may be to fix 2**J to be the size of the lowpass filter, period. Then the number of scales/wavelets/filters is adjusted so that they do not become wider than this.

This definition would not work in the `integral` mode, because there is no low-pass anymore. Maybe we can rephrase it as being the size of the receptive field, in the sense that two Diracs which are distant by more than `2**J` (in terms of Euclidean distance in time/pixel/voxel space) do not interfere in the scattering transform, or interfere negligibly.

**Comment by janden** [comment]
> > Against rejecting integers for 1D, though. I will make something that turns it into a tuple if it isn't one.
> 
> I agree. It would be good to have _some_ type dispatch in the constructor.

What's next, lists, NumPy arrays? Ok ok fine. I let my pedantry get ahead of me this time on this one.

> > One solution may be to fix 2**J to be the size of the lowpass filter, period. Then the number of scales/wavelets/filters is adjusted so that they do not become wider than this.
> 
> This definition would not work in the `integral` mode, because there is no low-pass anymore. Maybe we can rephrase it as being the size of the receptive field, in the sense that any two Diracs which are distant by more than `2**J` or more in the signal (in terms of Euclidean distance in time/pixel/voxel space) do not interfere in the scattering transform (or interfere negligibly because our Gabors do not have a compact support).

That's true. What you're suggesting makes more sense.

While we're on the subject, note that this definition is still awkward in the uber-vanilla case of dyadic wavelets (`Q=1`). For `J=2`, we have a lowpass filter is in the interval `[-pi/4, pi/4]`, the first wavelet in `[pi/2, pi]`, and a second wavelet in `[pi/4, pi/2]`. Makes sense, right? But now the (frequency) bandwidth of the lowpass filter is `pi/2` while the second wavelet has bandwidth `pi/4`, so we have a wavelet that's **wider** in time/space than the lowpass filter. In ScatNet this was _fixed_ by setting a special flag when `Q=1` but this is obviously not desirable.

**Comment by lostanlen** [comment]
Perfect. Thank you so much @eickenberg and @janden for review
